### PR TITLE
refactor: use interfaces and common utils for streaming destinations

### PR DESCRIPTION
**Pull request by koladilip**
# Description
    
Used Interfaces for better testability.
Added utils to handle AWS session creation.
Added basic IAM role support.

## Notion Ticket
https://www.notion.so/rudderstacks/AWS-FTR-for-streaming-destination-PoC-78b63869ac904ba49b0018a527ccb343

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

**Review by lvrach — COMMENTED** [review]
Very useful job here! It was high time to do it.
I left a few minor comments.

A small warning here, whenever we refactor this part of the code, they're some edge cases configuration in production that we end up breaking. For this I would suggest using this tool: 
https://github.com/rudderlabs/rudder-tests/tree/master/backtest-destinations

Further more, if we could introduce Ping methods for those destinations, so that we tests that all production configuration works without side-effects.

File: services/streammanager/common/common.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package common
```
[the remark this answers is not in the document]
Comment (koladilip) [reply services/streammanager/common/common.go:1]: It is added to avoid import cycle problem as streammanger package depends on destination specific managers so destination managers shouldn't import streammanager.

File: utils/googleutils/googleutils.go [hunk 2]
```diff
@@ -8,7 +8,7 @@ import (
 
 func CompatibleGoogleCredentialsJSON(jsonKey []byte) error {
 	if _, err := google.ConfigFromJSON(jsonKey); err == nil {
-		return fmt.Errorf("Google Developers Console client_credentials.json file is not supported")
+		return fmt.Errorf("google developers console client_credentials.json file is not supported")
```
[the remark this answers is not in the document]
Comment (koladilip) [reply utils/googleutils/googleutils.go:11]: We are constructing this custom error message when err := google.ConfigFromJSON(jsonKey), err is nil as we want to check if the user is using console credentials. If we get error means they are not using console credentials which is good for us.

File: services/streammanager/eventbridge/eventbridgemanager.go [hunk 3]
```diff
@@ -84,21 +55,16 @@ func Produce(jsonData json.RawMessage, producer, destConfig interface{}) (int, s
...
 
-		// fetching status code from response
-		if reqErr, ok := err.(awserr.RequestFailure); ok {
-			statusCode = reqErr.StatusCode()
-		}
-
-		return statusCode, err.Error(), err.Error()
+	if err != nil {
+		statusCode, respStatus, responseMessage := common.ParseAWSError(err)
+		pkgLogger.Errorf("[EventBridge] error  :: %s : %s", respStatus, responseMessage)
```
Comment (saikumarrs) [inline services/streammanager/eventbridge/eventbridgemanager.go:66]: Can we log the status code too here?

File: services/streammanager/firehose/firehosemanager.go [hunk 4]
```diff
@@ -1,159 +1,86 @@
...
-	err = json.Unmarshal(jsonConfig, &config)
-	if err != nil {
-		respStatus = "Failure"
-		responseMessage = "[FireHose] error  :: " + err.Error()
-		pkgLogger.Errorf("[FireHose] error  :: %v", err)
-		statusCode := 400
-		return statusCode, respStatus, responseMessage
+	client := producer.client
+	if client == nil {
+		return 400, "Failure", "[FireHose] error :: Could not create producer"
```
Comment (saikumarrs) [inline services/streammanager/firehose/firehosemanager.go:45]: Wouldn't it be better to validate this at the "NewProducer" stage?

**Review by saikumarrs — APPROVED** [review]

File: services/streammanager/common/common.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+func mapErrorMessageToStatusCode(errorMessage string, defaultStatusCode int) int {
+
+	if strings.Contains(errorMessage, "Throttling") {
+		// aws returns  "ThrottlingException"
+		// for throttling requests server will retry
+		return 429
+	}
+	if strings.Contains(errorMessage, " RequestExpired") {
```
[the remark this answers is not in the document]
Comment (koladilip) [reply services/streammanager/common/common.go:29]: Sure, will fix this

File: services/streammanager/eventbridge/eventbridgemanager.go (hunk 3 above)
Comment (koladilip) [reply services/streammanager/eventbridge/eventbridgemanager.go:65]: Sure, will fix this

File: services/streammanager/firehose/firehosemanager.go (hunk 4 above)
Comment (koladilip) [reply services/streammanager/firehose/firehosemanager.go:45]: We can but it is not necessary that we only use NewProducer to create the producer right, especially for testing purposes we pass our own mock client so to catch that this check is added.

Comment (saikumarrs) [reply services/streammanager/firehose/firehosemanager.go:45]: Makes sense.

File: services/streammanager/kafka/kafkamanager.go [hunk 6]
```diff
@@ -506,17 +507,17 @@ func prepareBatchOfMessages(topic string, batch []map[string]interface{}, timest
 }
 
 // CloseProducer closes a given producer
-func CloseProducer(ctx context.Context, pi interface{}) error {
+func (producer *KafkaProducer) CloseProducer() error {
```
[the remark this answers is not in the document]
Comment (koladilip) [reply services/streammanager/kafka/kafkamanager.go]: Sure

File: utils/awsutils/session.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+		},
+		Region:      aws.String(config.Region),
+		Credentials: createCredentails(config),
+	}))
+}
+
+func NewSessionConfig(destinationConfig interface{}, timeout time.Duration, serviceName string) (*SessionConfig, error) {
+	sessionConfig := SessionConfig{}
+
+	jsonConfig, err := json.Marshal(destinationConfig)
```
Comment (cisse21) [inline utils/awsutils/session.go:70]: Any reason why we would need to do both json marshal and unmarshal instead of just type checking the input interface?

Comment (koladilip) [reply utils/awsutils/session.go:70]: Downstream components is sending destinationConfig as interface{} and it cannot be type casted to SessionConfig but it has same fields as session config so marshalling to json and then unmarshalling to SessionConfig, actually this is an existing code just moved here. Ideal would be taking destination config as map[string]interface{} and manually extract required fields but this itself is bit long change as we need to write custom builders for each time from map type so currently not planning to address as part of this PR.

**Comment by koladilip** [comment]
@lvrach I have compared no difference in both iterations. I have files with me, I will share in our internal slack channel.

**Review by cisse21 — APPROVED** [review]

**Review by lvrach — APPROVED** [review]
